### PR TITLE
RSDK-7008: lock module manager while adding in parallel

### PR DESCRIPTION
**Pull request by maximpertsov**
Follow-up to https://github.com/viamrobotics/rdk/pull/3641 - we [stopped locking on individual module adds](https://github.com/viamrobotics/rdk/pull/3641/files#diff-a4308b77f384da81096eda21bf023471e80818dda86e74f9335f99a6a57d07c9L148-L149) to enable parallel adding, but there's no reason we can't lock while adding a batch.

Also adds:
* A new test module that isn't simple or complex module, to support testing scenarios that involve two modules
* Improves the `BuildTempModule` testing helper

### CODE REVIEW

File: module/modmanager/manager.go
Comment (benjirewis) [inline module/modmanager/manager.go:200]: Would you mind walking me through how this doesn't introduce deadlock with any attempted restarts offline?

Comment (maximpertsov) [reply module/modmanager/manager.go:200]: my understanding is that this would behave similarly to when `Add` was sequential - in that version of module manager, `Add` and `attemptRestart` both grabbed the manager lock without deadlocking - `Add` was able to finish and `attemptRestart` would run afterwards.

the difference now is that multiple `attemptRestart`s run if multiple modules crash while `Add`ing. my understanding is that this is fine - the `attemptRestart`s would run sequentially in some order after the batch `Add` is complete. each module has an `inRecovery` lock that prevents multiple `attemptedRestart`s from queuing up for the same module, but it's okay for restarts to queue up for different modules.

so the flow would be something like:

* `Add(mod1, mod2, mod3)`
* `mod1` succeeds but `mod2` and `mod3` crash
* `mod2` and `mod3` call their `attemptedRestart` functions, which idle until `mod1` is added
* the two `attemptedRestart` functions will be executed in some order sequentially

all this said, it's worth adding a testing that demonstrates the flow I am describing - i.e concurrently add two modules that both crash and show that there is no deadlock. I'll add that as part of this PR.

Comment (benjirewis) [reply module/modmanager/manager.go:200]: > the difference now is that multiple attemptRestarts run if multiple modules crash while Adding. my understanding is that this is fine - the attemptRestarts would run sequentially in some order after the batch Add is complete. each module has an inRecovery lock that prevents multiple attemptedRestarts from queuing up for the same module, but it's okay for restarts to queue up for different modules.

Also my understanding from reading the code, and thank you for writing it out! I think adding a test as you've described and running it with `-race` should be enough to say that this `Lock` you're adding is not _creating_ any deadlock or races.

Comment (maximpertsov) [reply module/modmanager/manager.go:200]: ah, so [I added a test for this scenario](https://github.com/viamrobotics/rdk/pull/3701/commits/39f249a3edd0cb5057243741367da2a99af18a4d) and do see a race condition! something about my assumptions we off (or I wrote the test poorly)

File: examples/customresources/demos/simplemodule/module.go
Comment (maximpertsov) [inline examples/customresources/demos/simplemodule/module.go:121]: reviewers, let me know if you don't think simplemodule should include these commands. if so I can make another testmodule for these multiple module scenarios.

Comment (benjirewis) [reply examples/customresources/demos/simplemodule/module.go:121]: [testmodule](https://github.com/viamrobotics/rdk/blob/main/module/testmodule/main.go#L108) should already have these commands; could we use that instead? The simple module should probably remain a "simple" example of a golang module.

Comment (benjirewis) [reply examples/customresources/demos/simplemodule/module.go:121]: Ah I see you're already using testmodule; was there a reason you also needed simple module to have the same functionality?

Comment (maximpertsov) [reply examples/customresources/demos/simplemodule/module.go:121]: I got an error related to adding duplicate resource namespaces if I tried to use testmodule twice

**Comment by maximpertsov** [comment]
@benjirewis looks like one more targeted lock gets rid of the race: https://github.com/viamrobotics/rdk/pull/3701/commits/b4d5356cbd3c37125d420ca9ba315ef8c8a05022

File: testutils/file_utils.go
Comment (maximpertsov) [inline testutils/file_utils.go:35]: I'm gonna go ahead and do this as part of this PR, but might wait for an implicit approval for the rest of the change, only because this might be end up being noisy.

I'm basically continuing to propagate [this pattern](https://sourcegraph.com/blog/go/advanced-testing-in-go#test-helpers) as I have for other test helpers.

Comment (maximpertsov) [reply testutils/file_utils.go:35]: here's [the resulting diff](https://github.com/maximpertsov/rdk/compare/RSDK-6770-lock-parallel-mod...maximpertsov:rdk:RSDK-6770-lock-parallel-mod-test-helper?expand=1)

File: examples/customresources/demos/simplemodule/module.go
Comment (maximpertsov) [reply examples/customresources/demos/simplemodule/module.go:121]: looks like one more targeted lock gets rid of the race: https://github.com/viamrobotics/rdk/pull/3701/commits/b4d5356cbd3c37125d420ca9ba315ef8c8a05022

File: testutils/file_utils.go
Comment (benjirewis) [reply testutils/file_utils.go:35]: That seems fine to me! I assumed you'll just modify `BuildTempModule` above? Do we still want to mark it `Helper` if we're actively failing inside it (I kind of forget how `Helper` works)?

File: module/modmanager/manager_test.go
Comment (benjirewis) [inline module/modmanager/manager_test.go:950]: Remove? I don't think you need to check `echo` functionality again.

File: module/modmanager/manager.go
Comment (benjirewis) [inline module/modmanager/manager.go:752]: Yep I like it! Not really sure why I used the non-locking version here. So this change gets rid of the data race you were seeing that also existed on `main`?

File: examples/customresources/demos/simplemodule/module.go
Comment (benjirewis) [reply examples/customresources/demos/simplemodule/module.go:121]: I'm still pretty against modifying the simple module example to have the echo and kill_module functionality; let me think about the best way to handle starting two modules that have the correct functionality 🤔 .

File: testutils/file_utils.go
Comment (maximpertsov) [reply testutils/file_utils.go:35]: Yes I was planning to commit the linked diff which replaces `BuildTempModule` with my `...New` definition.

We still want `t.Helper()` in this case I think - it makes so the test runner reports cleaner by reporting the line where the helper was called vs the line inside the helper whenever we have a failure. [This article](https://www.alexedwards.net/blog/easy-test-assertions-with-go-generics) had a quote that spells it out a little better than the golang docs:

> The [t.Helper()](https://pkg.go.dev/testing#T.Helper) function indicates to the Go test runner that our Equal() function is a test helper. This means that when t.Errorf() is called from our Equal() function, the Go test runner will report the filename and line number of the code which called our Equal() function in the output.

File: module/modmanager/manager_test.go
Comment (maximpertsov) [reply module/modmanager/manager_test.go:970]: man... such a sloppy copypasta on my part :disappointed:

File: examples/customresources/demos/simplemodule/module.go
Comment (maximpertsov) [reply examples/customresources/demos/simplemodule/module.go:121]: I agree with you, adding more functionality to a "simple" example feels like the opposite of we should be doing :slightly_smiling_face: 

If you can't think of a simple way to improve this, what do you think of me just copying our existing test module and creating a follow-up ticket to enable developers to create test modules as needed?

Comment (benjirewis) [reply examples/customresources/demos/simplemodule/module.go:121]: > If you can't think of a simple way to improve this, what do you think of me just copying our existing test module and creating a follow-up ticket to enable developers to create test modules as needed?

That SGTM. I think copying the existing test module and changing the API or model of the helper component should work. I can't think of a super easy to way to write a test utility that "creates" a whole module with the desired functionality, but if that doesn't sound infeasible to you, definitely throw it in a ticket!

Comment (maximpertsov) [reply examples/customresources/demos/simplemodule/module.go:121]: > I can't think of a super easy to way to write a test utility that "creates" a whole module with the desired functionality, but if that doesn't sound infeasible to you, definitely throw it in a ticket!

Agreed, all the paths I'm considering involve a decent amount of complexity, so best to leave as a follow-up IMO

Comment (maximpertsov) [reply examples/customresources/demos/simplemodule/module.go]: done https://github.com/viamrobotics/rdk/pull/3701/commits/d58d8cb3e1e96b31219b9320b56c611de0e6b23b

**Review by benjirewis — APPROVED** [review]
Generally looks good to me! Thanks for discovering the existing data race and fixing it 🧑‍🔧 .

**Comment by maximpertsov** [comment]
> Generally looks good to me! Thanks for discovering the existing data race and fixing it 🧑‍🔧 .

Thank **you** for encouraging me to investigate the flow a bit more :pray:

**Pull request merged**
